openfeature/client.py

```diff
@@ -464,3 +465,353 @@ def _typecheck_flag_value(value: typing.Any, flag_type: FlagType) -> None:
+                flag_evaluation,
+                reversed_merged_hooks,
+                hook_hints,
+            )
+
+            return flag_evaluation
+
+        except OpenFeatureError as err:
+            error_hooks(flag_type, hook_context, err, reversed_merged_hooks, hook_hints)
+
+            return FlagEvaluationDetails(
+                flag_key=flag_key,
+                value=default_value,
+                reason=Reason.ERROR,
+                error_code=err.error_code,
+                error_message=err.error_message,
+            )
+        # Catch any type of exception here since the user can provide any exception
+        # in the error hooks
+        except Exception as err:  # pragma: no cover
+            logger.exception(
+                "Unable to correctly evaluate flag with key: '%s'", flag_key
+            )
+
+            error_hooks(flag_type, hook_context, err, reversed_merged_hooks, hook_hints)
+
+            error_message = getattr(err, "error_message", str(err))
+            return FlagEvaluationDetails(
+                flag_key=flag_key,
+                value=default_value,
+                reason=Reason.ERROR,
+                error_code=ErrorCode.GENERAL,
+                error_message=error_message,
+            )
+
+        finally:
+            after_all_hooks(flag_type, hook_context, reversed_merged_hooks, hook_hints)
+
+    async def _create_provider_evaluation(
+        self,
+        provider: FeatureProvider,
+        flag_type: FlagType,
+        flag_key: str,
+        default_value: typing.Any,
+        evaluation_context: typing.Optional[EvaluationContext] = None,
+    ) -> FlagEvaluationDetails[typing.Any]:
+        """
+        Asynchronous encapsulated method to create a FlagEvaluationDetail from a specific provider.
+
+        :param flag_type: the type of the flag being returned
+        :param key: the string key of the selected flag
+        :param default_value: backup value returned if no result found by the provider
+        :param evaluation_context: Information for the purposes of flag evaluation
+        :return: a FlagEvaluationDetails object with the fully evaluated flag from a
+        provider
+        """
+        args = (
+            flag_key,
+            default_value,
+            evaluation_context,
+        )
+
+        get_details_callables: typing.Mapping[FlagType, GetDetailCallable] = {
+            FlagType.BOOLEAN: provider.resolve_boolean_details,
+            FlagType.INTEGER: provider.resolve_integer_details,
+            FlagType.FLOAT: provider.resolve_float_details,
+            FlagType.OBJECT: provider.resolve_object_details,
+            FlagType.STRING: provider.resolve_string_details,
+        }
+
+        get_details_callable = get_details_callables.get(flag_type)
+        if not get_details_callable:
+            raise GeneralError(error_message="Unknown flag type")
+
+        resolution = await get_details_callable(*args)  # type: ignore[misc]
```

Comment:
Might be helpful to check [whether `get_details_callable` is awaitable](https://docs.python.org/3/library/inspect.html#inspect.isawaitable) here in order to make sync provider compatible with async clients?

It might look like this, but I didn't test it
```suggestion
        resolution = get_details_callable(*args)  # type: ignore[misc]
        if isawaitable(resolution):
            resolution = await resolution
```

Reply:
I actually think this has to be done. If not I think we would get type errors right?